### ISSUE TITLE
Be able to customize the classpath vs. modulepath decision

### ISSUE THREAD

**Issue by iksnalybok** [gradle/gradle#13147]
Since 6.4, the decision to put a jar in the classpath or the modulepath is taken by the org.gradle._internal_.jvm.JavaModuleDetector. It applies IMHO a pretty nice heuristic, but is not flexible at all.

You may want to customize it. I've for example the case of a library with native code (so/dll), which has a module-info class. Such jars usually embed some code to automatically extract and load the so/dll files. It works nicely when the jar is in the classpath, but fail to "getResource" when being in the module path. I cannot force gradle to put it in the classpath. I failed to hack the JavaExec task. At some point, a --module-path option is appended and takes precedence. I ended up reusing my own pre-6.4 class/module paths splitter, and used it to generate shell scripts to do the job.

On the API level, why don't you offer both a classpath and a modulepath, like the java/javac cli does? The current api only offers a classpath (so we can hack it), which get arbitrarily split by the JavaModuleDetector. This would allow to hack the modulepath too.

On the build.gradle[.kts] side, we need something to avoid any hacking. Currently, I'm using dedicated (non-transitive) configurations to specify the dependencies to manage manually (not sure it's the best option though).

Is something planned to bring flexible support for classpath vs modulepath decision?

**Comment by ljacomet** [comment]
Relevant issues: #12630 

Question about that library with native code: Is there any plan to make the loading of the native code work on the module path?
If not, and if you know, why does it have a `module-info` at all?

**Comment by iksnalybok** [comment]
No idea. There is indeed something to fix at the library level.

My point here is that javac/java allow me to set both the classpath and the module path, and my app compiles and runs. Gradle does not allow that, so I cannot use it to compile and run (in my example the compilation works as native libs are runtimeOnly dependencies).

Thanks for the reference. I had went through it (before 6.4 was released), and the title is a bit misleading. I was then pleased to see 6.4 bring builtin support for modules. I understand from #12630 and the various modules-aware plugins that it's hard to find a proper way to handle modules. I personally would like to see some flexibility, i.e. let the build.gradle[.kts] be able to amend gradle default decision. There are also some module-related tooling, like jlink or jpackage, which are not integrated. The 'application' plugin does not deal with --add-opens, --add-reads, etc. options too. In the meantime, offering some insights on dependencies would ease custom integration of these tools.

But my question was more about gradle's vision. The 6.4 release notes [claims](https://docs.gradle.org/6.4/release-notes.html#java-modules) "_With this release, Gradle supports the Java Module System with everything you need to compile and execute tests for Java modules._" Does this means that jpms support is complete and we need to rely on 3rd party plugins for extra features? or is it just a first step, towards more advanced support?

**Comment by big-andy-coates** [comment]
+1, just ran into the same issue and ended up with "module not found" errors because a `JavaExec` task was putting the jar on the class path, rather than the module path.

**Comment by jfu334** [comment]
Same problem with org.fxmisc.richtext:richtextfx:0.10.9. 

It relies on org.reactfx:reactfx:2.0-M5. This is not a module, but richtextfx relies in the module name implicitly created for it. Gradle just puts it in classpath, not modulepath. 

As implicit modules are a thing, the feature in Gradle to automatically determine weather it's a module or not is fundamentally broken.

**Comment by Vampire** [comment]
Aaaactually, depending on implicit automatic modules is fundamentally broken and noone should do that, especially not in a library. :-)

**Comment by jfu334** [comment]
We're not here to judge the feature of implicit modules. It is specified and supported by java, but not by gradle. In a perfect world no one should have to rely on it of course. Still there are valid use cases for it when using legacy code.

**Comment by Vampire** [comment]
Sure it is supported by Gradle.
It just does what the Gradle folks think makes most sense as a convention by default as usual, but always provides the flexibility to do whatever you want.
So if you want to have it differently, just configure it however you need it.
You can for example set `inferModulePath` to false to disable the Gradle logic completely and do all decisions yourself.

**Comment by Vampire** [comment]
It has `modularity` which has `inferModulePath` as also documented in the user guide at https://docs.gradle.org/current/userguide/java_library_plugin.html#disabling_java_module_support :-)

To set the module path, you just use a standard `jvmArgs` call or a `jvmArgumentProvider` depending on what is more appropriate.

**Comment by jfu334** [comment]
After some intense fiddling around with JavaExec I've come to the conclusion that it is fundamentally broken since the module system was implemented in java. It makes easy tasks extremely complicated. A build system is supposed to do the opposite. The documentation is garbage. 

I've managed to get the command line nearly right with inferModulePath set to false, but now I can't get gradle to start the code as module. It insists of passing the mainClass to the java call (as in java -p ... -cp ... <mainClass>) instead of --module <mainModule>/<mainClass>. I don't see a way around this.

It's also not possible to manually supply these through jvmArgs, JavaExec doesn't work without a mainClass set and then always passes it to the java call. mainModule is always ignored, the doc does not state conditions under which it is used or ignored and I don't have time to look in the source code of it.

I will use the Exec task instead and build the command line manually. By spending some hours on getting JavaExec to work I've at least learned enough about the gradle internals to prepare the Exec call directly. (There were other things JavaExec did wrong, like passing the main code through -cp killing all resource loads.)

**Comment by Vampire** [comment]
Yeah, it is non-discussable that there is much room and need for improvement, but if you are desperate enough, you can at least make it work. :-D

`--module` is only used by `JavaExec` if `inferModulePath` is enabled and `mainModule` is set.
So if you really need to disable `inferModulePath` but still want to run your app on the module path, you need a little trickery like this:
```kts
application {
    mainModule = "org.example"
    mainClass = "org.example.Main"
}

tasks.run.configure {
    modularity.inferModulePath = false
    mainClass = "-Dfoo"
    /* Configure class and module path as you need them */
    argumentProviders.add(
        object : CommandLineArgumentProvider {
            @Input
            val mainModule = application.mainModule.zip(application.mainClass) { m, c -> "$m/$c" }

            override fun asArguments() = listOf(
                "--module",
                mainModule.get()
            )
        }
    )
}
```
Or you just do not disable `inferModulePath`.

**Comment by jfu334** [comment]
Setting mainClass to a JVM option looks funny, hadn't had this idea. It has the same grasp for quality as JavaExec itself, so I guess it's the right solution for the right problem. Maybe this should be included in the docs. 

It took me 10 minutes to get the Exec task running and fix the problems gradle introduces with the module. For those who stumble on this issue and search for a copy&paste solution:

```
task run(type:Exec)
{
    def modules=sourceSets.main.runtimeClasspath
        .minus(project.files([new File("${buildDir}/classes/java/main"), new File("${buildDir}/resources/main")]))
        .plus(project.files([jar.archivePath]))

    def entry="someModule/someClass"
    def jvmArgs=[lots of, '-p', modules.asPath]
    def runArgs=[lots of]

    executable 'java'
    args=jvmArgs+['--module', entry]+runArgs
}
```

Does anyone know why runtimeClasspath contains the compile output instead of the module jar? This also happened with inferModulePath enabled.

**Comment by Vampire** [comment]
> Does anyone know why runtimeClasspath contains the compile output instead of the module jar? This also happened with inferModulePath enabled.

Because Gradle is pretty good in avoiding unnecessary work.
And when not running on the module path, it is perfectly fine to just use the directories with the class files and resources directly without needlessly building a jar.
When run on the module path, this can of course lead to problems unless you also use `--patch-module` parameters to combine those directories into one module.

Due to that, the `application` plugin configures the `run` task in a way that when `mainModule` is set on it, it uses `jar.outputs.files + configurations.runtimeClasspath`, while it uses `sourceSets.main.runtimeClasspath` otherwise.

**Comment by jfu334** [comment]
But Gradle already builds the JAR and Gradle should know that it's a module as the JARs source set contains a module-info.java. In this situation providing sources/resources by -cp will definitely break things. 

Sounds like these issues were fixed in the application plugin but not in the JavaExec task. I don't want to build an application, the target is only for manual testing and contains very specific options. Of course I could simply create a run target in the IDE or use a script to run the thing, but that defeats the whole purpose of JavaExec. 

Maybe instead of trying to fix JavaExec with additional options that, in the end, don't do the job (like mainModule) it would be easier to provide a new Task like JavaModuleExec that is better suited to the case.

**Comment by Vampire** [comment]
> But Gradle already builds the JAR

Only when necessary.

> and Gradle should know that it's a module as the JARs source set contains a module-info.java.

Doesn't mean it has to be used. If you for example do white-box tests (non-modular test task), it might be perfectly fine to put it on the classpath. Especially with the unfortunate wrong premise Gradle currently follows as handled also in #25954.

> In this situation providing sources/resources by -cp will definitely break things.

It will not "definitely break things", it will only break if some module-path specific things are used, like new service lookup mechanism without supporting also running on the classpath by supporting the old mechanism.

> Sounds like these issues were fixed in the application plugin but not in the JavaExec task.

No, the `application` plugin adds and configures the `JavaExec` task.
Tasks and extensions are usually as opinionless as possible to improve reusability and plugins add the opinion by registering extensions and tasks, wiring them together, and setting default values.

> I don't want to build an application, the target is only for manual testing and contains very specific options. Of course I could simply create a run target in the IDE or use a script to run the thing, but that defeats the whole purpose of JavaExec.

Well, as I said, this is not about `JavaExec` which has no opinion.
It uses the classpath you give it, or the jvm arg for module path you give it.
If you give `sourceSets.main.runtimeClasspath` then that is the libs plus directories.
If you give `jar.outputs.files + configurations.runtimeClasspath` then that is only jars.

**Comment by jfu334** [comment]
This is all very pointless hairsplitting.

Your example (the -Dfoo option for mainClass) does demonstrate very well that the JavaExec Task isn't able to properly start the module Gradle has just build. This is nowhere documented. Indeed the docs suggest JavaExec could properly handle java modules but it can't. There should be a big warning on the doc page about this, right on the first line, maybe even in red letters.  

And to be clear on this, if you start a java module not as a module you're essentially doing everything wrong. Unless you're doing it for a very specific purpose (like testing or some ancient incompatible garbage) and know the consequences. 

For the ServiceLoader: Yes people use that feature. It's written in the official Java Specs for *years*. By incidence I'm actually using it in this very project. If JavaExec doesn't support it, it doesn't support proper Java. 

Also without a properly started module you mess up certain resource loads.

**Comment by Vampire** [comment]
> This is all very pointless hairsplitting.

I didn't split any hair. I just tried to answer your questions and help you make it do what you want.

> Your example (the -Dfoo option for mainClass) does demonstrate very well that the JavaExec Task isn't able to properly start the module Gradle has just build.

No, it does not, see below.

> Indeed the docs suggest JavaExec could properly handle java modules but it can't.

It can. If you don't disable it. If you disable it by disabling `inferModulePath` or not setting a `mainModule`, you need such dirty hacks like the `-Dfoo` to trick it nevertheless do what you want. That's why I said you should probably not disable `inferModulePath` , then you don't need hacks that dirty. Again, tasks are dumb, they just do what you configure them to do and if you configure it to be not JPMS-aware, then ... well ... it is not JPMS-aware.

> And to be clear on this, if you start a java module not as a module you're essentially doing everything wrong.

I fully agree. That's why the `application` plugin configures the `run` task it adds properly so that it is started properly.
If you register an own task of that type, it is your responsibility to configure it however you need it, either to run properly or for your specific purpose.

Please let's stop polluting this issue though and constantly notifying each watcher with only faintly related discussions.
If you want to continue discussing something, please carry it over to a more appropriate place like the community Slack, or the community forum.

**Comment by jfu334** [comment]
> Please let's stop polluting this issue though and constantly notifying each watcher with only faintly related discussions. If you want to continue discussing something, please carry it over to a more appropriate place like the community Slack, or the community forum.

You're right. Sorry if the last post came along rude. I have to use Gradle to make things work, diving into the (poorly documented) internals and even the source code to understand how it works can be quite frustrating. 

From what I've seen in the source the `application` plugin might suffer from the same issues as JavaExec (`inferModulePath` set to false won't start a module properly), but I haven't tested it. I couldn't find code to mitigate the issue of JavaExec. 

Also it is not clear weather JavaExec is intended for regular use by the user or an implementation detail to be used by other Tasks/Plugins. This decision leads to different requirements regarding quality, stability and documentation.
But to my feeling that's a Problem throughout Gradle in general.
